File: apps/backend/app/schemas/lease.py
from datetime import date, datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, EmailStr, Field


# ── Discount schemas ──────────────────────────────────────────────────────────

class LeaseDiscountCreateRequest(BaseModel):
    label: str
    type: str  # "fixed" | "percentage"
    value: float = Field(gt=0)
    effective_from: date
    effective_to: Optional[date] = None
    note: Optional[str] = None


class LeaseDiscountResponse(BaseModel):
    id: str
    label: str
    type: str
    value: float
    effective_from: date
    effective_to: Optional[date] = None
    note: Optional[str] = None
    recorded_by: str
    created_at: datetime
    # computed fields
    effective_rent: float   # rent after this discount applied
    discount_amount: float  # KES saving


# ── Escalation schemas ────────────────────────────────────────────────────────

class RentEscalationCreateRequest(BaseModel):
    effective_date: date
    new_rent_amount: float = Field(gt=0)
    note: Optional[str] = None


class RentEscalationResponse(BaseModel):
    id: str
    effective_date: date
    new_rent_amount: float
    percentage_increase: Optional[float]
    applied: bool
    applied_at: Optional[datetime]
    note: Optional[str]
    created_by: str
    created_at: datetime


# ── Early termination schemas ─────────────────────────────────────────────────

class EarlyTerminationTermsRequest(BaseModel):
    penalty_type: str = "months"
    penalty_value: float = Field(gt=0)
    notice_days: int = Field(ge=0, default=30)
    note: Optional[str] = None


class EarlyTerminationTermsResponse(BaseModel):
    penalty_type: str
    penalty_value: float
    notice_days: int
    note: Optional[str]
    penalty_amount: float  # computed: months*rent or fixed


# ── Renewal offer schemas ─────────────────────────────────────────────────────

class RenewalOfferCreateRequest(BaseModel):
    new_rent_amount: float = Field(gt=0)
    new_end_date: Optional[date] = None
    message: Optional[str] = None


class RenewalOfferResponse(BaseModel):
    id: str
    new_rent_amount: float
    new_end_date: Optional[date]
    message: Optional[str]
    status: str
    sent_at: datetime
    responded_at: Optional[datetime]
    created_by: str


# ── Co-tenant / guarantor schemas ─────────────────────────────────────────────

class CoTenantCreateRequest(BaseModel):
    role: str = "co_tenant"
    first_name: str
    last_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    id_type: Optional[str] = None
    id_number: Optional[str] = None


class CoTenantResponse(BaseModel):
    id: str
    role: str
    first_name: str
    last_name: str
    email: Optional[str]
    phone: Optional[str]
    id_type: Optional[str]
    id_number: Optional[str]
    added_at: datetime
    added_by: str


# ── Notes & rating schemas ────────────────────────────────────────────────────

class LeaseNoteCreateRequest(BaseModel):
    body: str
    is_private: bool = True


class LeaseNoteResponse(BaseModel):
    id: str
    body: str
    is_private: bool
    created_by: str
    created_at: datetime


class TenantRatingRequest(BaseModel):
    score: int = Field(ge=1, le=5)
    payment_timeliness: int = Field(ge=1, le=5, default=3)
    property_care: int = Field(ge=1, le=5, default=3)
    communication: int = Field(ge=1, le=5, default=3)
    note: Optional[str] = None


class TenantRatingResponse(BaseModel):
    score: int
    payment_timeliness: int
    property_care: int
    communication: int
    note: Optional[str]
    rated_by: str
    rated_at: datetime


# ── Inline tenant creation ────────────────────────────────────────────────────

class TenantCreateInline(BaseModel):
    """Create a new tenant account atomically when creating a lease."""
    email: EmailStr
    first_name: str
    last_name: str
    phone: Optional[str] = None
    password: str = Field(min_length=8)


# ── Lease CRUD ────────────────────────────────────────────────────────────────

class LeaseCreateRequest(BaseModel):
    unit_id: str
    # provide either tenant_id (existing) or tenant_create (new)
    tenant_id: Optional[str] = None
    tenant_create: Optional[TenantCreateInline] = None
    onboarding_id: Optional[str] = None
    start_date: date
    end_date: Optional[date] = None
    rent_amount: float = Field(gt=0)
    deposit_amount: float = Field(ge=0)
    utility_deposit: Optional[float] = Field(default=None, ge=0)
    notes: Optional[str] = None


class LeaseResponse(BaseModel):
    id: str
    reference_no: str
    org_id: str
    property_id: str
    unit_id: str
    unit_code: Optional[str] = None           # denormalised from Unit
    tenant_id: str
    onboarding_id: Optional[str]
    onboarding_token: Optional[str] = None    # present only on create; used to redirect to /onboarding/<token>
    status: str
    start_date: date
    end_date: Optional[date]
    rent_amount: float
    deposit_amount: float
    utility_deposit: Optional[float]
    unit_utility_deposits: float = 0.0        # sum of utility_overrides[*].deposit
    notes: Optional[str]
    signed_at: Optional[datetime]
    activated_at: Optional[datetime]
    terminated_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    discounts: List[LeaseDiscountResponse] = []
    effective_rent: float = 0.0               # rent after active discounts
    discount_amount: float = 0.0              # total KES discount currently active
    escalations: List[RentEscalationResponse] = []
    early_termination: Optional[EarlyTerminationTermsResponse] = None
    renewal_offer: Optional[RenewalOfferResponse] = None
    co_tenants: List[CoTenantResponse] = []
    notes_internal: List[LeaseNoteResponse] = []
    rating: Optional[TenantRatingResponse] = None


class LeaseListResponse(BaseModel):
    items: List[LeaseResponse]
    total: int
    page: int
    page_size: int


# ── Unit pricing breakdown ────────────────────────────────────────────────────

class UtilityLineItem(BaseModel):
    key: str
    label: str
    type: str                   # shared | metered | subscription
    rate: Optional[float]
    unit_label: Optional[str]   # e.g. KWh, m³, KES/mo
    income_account: Optional[str]
    deposit: Optional[float] = None  # per-utility one-time deposit


class UnitPricingResponse(BaseModel):
    unit_id: str
    unit_code: str
    rent_amount: float
    deposit_amount: float
    deposit_rule: str
    utility_deposit: Optional[float]
    utilities: List[UtilityLineItem]
    prorated_rent: float            # rent for remaining days of move-in month
    prorated_days: int              # number of days charged
    days_in_month: int              # total days in move-in month
    total_move_in: float            # deposit + utility_deposit + prorated_rent

File: apps/frontend/src/types/lease.ts
export type LeaseStatus = 'draft' | 'pending_payment' | 'pending_signature' | 'active' | 'expired' | 'terminated'

export interface LeaseDiscount {
  id: string
  label: string
  type: 'fixed' | 'percentage'
  value: number
  effective_from: string
  effective_to?: string
  note?: string
  recorded_by: string
  created_at: string
  effective_rent: number
  discount_amount: number
}

export interface LeaseDiscountPayload {
  label: string
  type: 'fixed' | 'percentage'
  value: number
  effective_from: string
  effective_to?: string
  note?: string
}

export interface RentEscalation {
  id: string
  effective_date: string
  new_rent_amount: number
  percentage_increase?: number
  applied: boolean
  applied_at?: string
  note?: string
  created_by: string
  created_at: string
}

export interface EarlyTerminationTerms {
  penalty_type: 'months' | 'fixed'
  penalty_value: number
  notice_days: number
  note?: string
  penalty_amount: number
}

export interface RenewalOffer {
  id: string
  new_rent_amount: number
  new_end_date?: string
  message?: string
  status: 'pending' | 'accepted' | 'declined' | 'expired'
  sent_at: string
  responded_at?: string
  created_by: string
}

export interface CoTenant {
  id: string
  role: 'co_tenant' | 'guarantor'
  first_name: string
  last_name: string
  email?: string
  phone?: string
  id_type?: string
  id_number?: string
  added_at: string
  added_by: string
}

export interface LeaseNote {
  id: string
  body: string
  is_private: boolean
  created_by: string
  created_at: string
}

export interface TenantRating {
  score: number
  payment_timeliness: number
  property_care: number
  communication: number
  note?: string
  rated_by: string
  rated_at: string
}

export interface TenantCreateInline {
  email: string
  first_name: string
  last_name: string
  phone?: string
  password: string
}

export interface Lease {
  id: string
  reference_no: string
  org_id: string
  property_id: string
  unit_id: string
  unit_code?: string              // denormalised from Unit
  tenant_id: string
  onboarding_id?: string
  onboarding_token?: string       // present only on create response
  status: LeaseStatus
  start_date: string
  end_date?: string
  rent_amount: number
  deposit_amount: number
  utility_deposit?: number
  unit_utility_deposits: number   // sum of utility_overrides[*].deposit
  notes?: string
  signed_at?: string
  activated_at?: string
  terminated_at?: string
  created_at: string
  updated_at: string
  discounts: LeaseDiscount[]
  effective_rent: number
  discount_amount: number
  escalations: RentEscalation[]
  early_termination?: EarlyTerminationTerms
  renewal_offer?: RenewalOffer
  co_tenants: CoTenant[]
  notes_internal: LeaseNote[]
  rating?: TenantRating
}

export interface LeaseCreateRequest {
  unit_id: string
  tenant_id?: string
  tenant_create?: TenantCreateInline
  onboarding_id?: string
  start_date: string
  end_date?: string
  rent_amount: number
  deposit_amount: number
  utility_deposit?: number
  notes?: string
}
